### PR TITLE
vim: work around Xcode 16 bug

### LINKED ISSUE #195325: Vim bottle is miscompiled because Xcode 16 has codegen bugs
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
HOMEBREW_VERSION: 4.4.2-23-g62c1f5b
ORIGIN: https://github.com/Homebrew/brew
HEAD: 62c1f5b42bb9959414d1e016f96385f48b9b961d
Last commit: 77 minutes ago
Core tap JSON: 23 Oct 08:36 UTC
Core cask tap JSON: 23 Oct 08:36 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.UhYxpJnKWC/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.5 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.6.1-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: 16.0
Rosetta 2: false

Your system is ready to brew
```

### Verification

- [X] My `brew doctor` output says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

The current bottled Vim seems to have behavior bugs because the compiler is miscompiling it. From what we can tell Xcode 16 seems to have a fair amount of codegen bugs leading to this.

This issue was discovered in https://github.com/vim/vim/pull/15764, when we attempted to upgrade Vim's CI to use macos-15 runners. It failed one of the tests, and after investigation it appears to be an inherent Xcode 16 clang bug that has not yet been fixed in Xcode.

Note that Vim doesn't seem to be the only affected software. From browsing around it appears JDK is also facing issues with Xcode 16 clang generating wrong results (https://bugs.openjdk.org/browse/JDK-8340341).

I tried to look around but I couldn't find other reports in Homebrew. I'm guessing this is either an niche bug affecting only few software, or the bug is so subtle that it is not easy to detect. Are all bottles in Homebrew built with Xcode 16?

Edit: Note that this affects Neovim as well. One caveat is that Vim bottles seem to be broken in both macOS 14/15 but Neovim bottles seem to be broken only in macOS 15. Using `otool -l` and looking under `LC_BUILD_VERSION`, it does seem that Vim is built with Xcode 16 for both macOS 14/15, and Neovim is only using Xcode 16 to build bottles for macOS 15, so that matches the observed behaviors.

### What happened (include all command output)?

The cursor landed on the `<82>` part, which is not the expected behavior. This doesn't happen when Vim is built in Xcode 15 or when building in Xcode 16 using `-O1`/`-O0`.

### What did you expect to happen?

The cursor should land on the `<e2>` part, which is where the invalid UTF-8 sequence starts. This is how it works on macOS 14.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. `brew install vim` (this will download the v9.1.0800 bottled version of Vim as of this writing)
2. `vim`
3. Enter this command in Vim `:call setline(1, "abcdef\xE2\x82xyz")`.
4. Type `8g8` (this Vim command finds the first invalid UTF-8 sequence).
5. Observe
```

### PR DESCRIPTION
**Pull request by fxcoudert**
Fixes https://github.com/Homebrew/homebrew-core/issues/195325 (hopefully)

Can someone try the fix to confirm it works?

### CODE REVIEW

File: Formula/v/vim.rb [hunk 1]
```diff
@@ -61,6 +61,10 @@ def install
     # vim doesn't require any Python package, unset PYTHONPATH.
     ENV.delete("PYTHONPATH")
 
+    if DevelopmentTools.clang_build_version == 1600
+      ENV.append "CFLAGS", "--with-extra-cflags=-mllvm -enable-constraint-elimination=0"
```
Comment (carlocab) [inline Formula/v/vim.rb:65]:
```suggestion
      ENV.append_to_cflags "-mllvm -enable-constraint-elimination=0"
```
`--with-extra-cflags` is an OpenJDK `configure` argument.

Comment (fxcoudert) [reply Formula/v/vim.rb:65]: Yeah, copy-pasto :( thanks

File: Formula/v/vim.rb [hunk 2]
```diff
@@ -61,6 +61,10 @@ def install
     # vim doesn't require any Python package, unset PYTHONPATH.
     ENV.delete("PYTHONPATH")
 
+    if DevelopmentTools.clang_build_version == 1600
+      ENV.append_to_cflags "-mllvm -enable-constraint-elimination=0"
+    end
```
Comment (carlocab) [inline Formula/v/vim.rb:66]:
```suggestion
    ENV.append_to_cflags "-mllvm -enable-constraint-elimination=0" if DevelopmentTools.clang_build_version == 1600
```
Should fix your syntax failure.

**Comment by carlocab** [comment]
For the record, I did try this and it seems to fix the reproducer at #195325.

**Review by p-linnane — APPROVED** [review]

**Comment by p-linnane** [comment]
Worked for me locally. Probably want to rev bump as well since the current bottle is broken?

**Review by carlocab — APPROVED** [review]

**Comment by fxcoudert** [comment]
> Probably want to rev bump as well since the current bottle is broken?

Broken for a very specific use case only. I thought it was not worth it.

**Comment by ychin** [comment]
Thanks for the fix!

Since we have identified the exact issue (when I filed the original issue we hadn't identified a flag that could fix it yet) I think it could be worth it to just add this to Vim upstream to auto-detect the issue and add the build flags since this issue happens regardless of it being built in Homebrew or not. If we end up doing that I'll file a PR to revert the Homebrew change here although hopefully Xcode (16.2 hopefully) would have fixed it.

**Comment by calvinit** [comment]
#194719

**Comment by ychin** [comment]
Should this be reverted now that Xcode 16.2 is out fully, and has fixed the issue?

**Comment by fxcoudert** [comment]
We will remove it when our CI machines are updated to Xcode 16.2, which is not yet the case.

**Pull request merged**